proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,260 @@
+
+We propose to also store the new WALs in separate directories e.g. `wal.v2`. Thanks to that the rewrite from one version to another is **eventual** and can be done segment by segment without any forcible rewrite. The WAL will be rewritten within the next 2h of a normal operation. The additional advantage of this way of versioning is that it's clear when your WAL fully migrated to a certain version.
+
+Finally, we propose that all Prometheus releases will contain the following table:
+
+| Data        | Supported | Writes |
+|-------------|-----------|--------|
+| WAL         | v1, v2    | v1     |
+| Block index | v2        | v2     |
+
+```
+The last revertable Prometheus version: v3.0 
+```
+
+See [alternatives](#alternatives) for other ideas.
+
+### Two-Fold Migration Strategy
+
+Given the following example:
+
+![twofold.png](../assets/2024-11-25_changing_wal_format/twofold.png)
+
+1. We release Prometheus X+1 version that supports both Y and Y+1 data but still writes Y.
+2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y.
+
+While this example shows only one version where of forward compatibility (when Y and Y+1 are supported, but Y is still written), in practice there could be more "forward compatible releases" within this strategy.
+
+Technically we have to keep old format read (and write) support until the next major release.
+Pros:
+* Users have a durable rollout path back and forth.
+* Dev has clarity on how to develop "breaking revert" changes.
+* In theory, it allows revert to X from X+2, by going through X+1 and ensuring all data was migrated (eventually) to Y version. In practice however that eventuality is long, or hard to discover.
+
+Cons:
+* It can catch users by surprise.
+  * **Mitigation**: We also plan guide and documentation.
+* It takes time to rollout new changes if you want them fast.
+  * **Mitigation**: We plan to add a flag to opt-in sooner.
+
+## Alternatives
+
+### Require LTS support
+
+We could add a variation to the [Two-Fold Migration Strategy](#two-fold-migration-strategy) (let's call it a "LTS migration strategy") where both X+1 and the last LTS (long time support version) before X+1, is able to read Y+1 version. Only then we are allowed to release X+2 that switches the default.
+
+For example:
+
+1. LTS 3.1 only supports WAL v1.
+2. 3.3 adds WAL v2.
+3. we wait unit next LTS so e.g. 3.24.
+4. 3.25 can now switch to WAL v2.
+5. 4.0 can remove WAL v1 support.
+
+Pros:
+* Gives a bit more stability to users and less surprises.
+
+Cons:
+* Extremely heavy process that will make us afraid/refuse to make improvements to WAL, because it's too much work. It might fails our goal of `Balancing development velocity with user data stability risks`
```

Comment:
```suggestion
* Extremely heavy process that will make us afraid/refuse to make improvements to WAL, because it's too much work. It might fail our goal of `Balancing development velocity with user data stability risks`
```